deputy.go

```diff
@@ -0,0 +1,182 @@
+// Copyright 2015 Canonical Ltd.
+// Licensed under the LGPLv3, see LICENCE file for details.
+
+// Package deputy provides more advanced options for running commands.
+package deputy
+
+import (
+	"bufio"
+	"bytes"
+	"fmt"
+	"io"
+	"os/exec"
+	"strings"
+	"time"
+)
+
+// ErrorHandling is a flag that tells Deputy how to handle errors running a
+// command.  See the values below for the different modes.
+type ErrorHandling int
+
+const (
+	// DefaultErrs represents the default handling of command errors - this
+	// simply returns the error from Cmd.Run()
+	DefaultErrs ErrorHandling = iota
+
+	// FromStderr tells Deputy to convert the stderr output of a command into
+	// the text of an error, if the command exits with an error.
+	FromStderr
+
+	// FromStdout tells Deputy to convert the stdout output of a command into
+	// the text of an error, if the command exits with an error.
+	FromStdout
+)
+
+// Deputy is a type that runs Commands with advanced options not available from
+// os/exec.  See the comments on field values for details.
+type Deputy struct {
+	// Timeout represents the longest time the command will be allowed to run
+	// before being killed.
+	Timeout time.Duration
+	// Errors describes how errors should be handled.
+	Errors ErrorHandling
+	// StdoutLog takes a function that will receive lines written to stdout from
+	// the command.
+	StdoutLog func(string)
+	// StdoutLog takes a function that will receive lines written to stderr from
+	// the command.
+	StderrLog func(string)
```

Comment:
I wonder if it would be better for these to take a []byte argument to save an allocation,
which can be reasonably significant for verbose logging.